benchmark/atoi.cpp

```diff
@@ -173,11 +168,33 @@ auto leadingDigitsCount(const char *p) noexcept {
     }
 }
 
-int c_strToI(const char *str) noexcept {
-    constexpr static std::array<int, 8> LastFactor = {
-        1, 10, 100, 1000,
-        10'000, 100'000, 1000'000, 10'000'000
-    };
+namespace impl {
+
+template<typename> struct ConversionTraits;
+template<> struct ConversionTraits<int32_t>{
+    constexpr static auto NPositions = 9; // from 10^0 to 10^8
+    using PowersOf10Array = std::array<int32_t, NPositions>;
+    using DoublePrecision = uint64_t;
+};
+template<> struct ConversionTraits<int64_t>{
+    constexpr static auto NPositions = 17; // from 10^0 to 10^16
+    using PowersOf10Array = std::array<int64_t, NPositions>;
+    using DoublePrecision = __uint128_t;
+};
+
+template<typename Result>
+auto PowersOf10Array() {
+    using Traits = ConversionTraits<Result>;
+    typename Traits::PowersOf10Array rv{1};
+    for (std::size_t i = 1; i < Traits::NPositions; ++i) {
+        rv[i] = rv[i - 1] * 10;
+    }
+    return rv;
+};
+
+template<typename Return>
+Return c_strToIntegral(const char *str) noexcept {
```

Comment:
A test for the version that handles 16 byte chunks that crosses a 16 byte boundary seems wise, but perhaps I missed it.